diff --git a/lib/compat.h b/lib/compat.h
index a5cf7280..b4d98e4c 100644
--- a/lib/compat.h
+++ b/lib/compat.h
@@ -143,7 +143,9 @@ typedef char bool;
 #define ZIP_OFF_MAX ZIP_INT16_MAX
 #define ZIP_OFF_MIN ZIP_INT16_MIN
 #else
-#error unsupported size of off_t
+//#error unsupported size of off_t
+#define ZIP_OFF_MAX ZIP_INT32_MAX
+#define ZIP_OFF_MIN ZIP_INT32_MIN
 #endif
 
 #if defined(HAVE_FTELLO) && defined(HAVE_FSEEKO)